src/connectors/query-rules/connectQueryRules.ts

```diff
@@ -33,15 +49,175 @@ export type QueryRulesConnector = <T>(
   unmount?: () => void
 ) => QueryRulesWidgetFactory<T>;
 
+const withUsage = createDocumentationMessageGenerator({
+  name: 'query-rules',
+  connector: true,
+});
+
+function hasStateRefinements({
+  disjunctiveFacetsRefinements,
+  facetsRefinements,
+  hierarchicalFacetsRefinements,
+  numericRefinements,
+}) {
+  return [
+    disjunctiveFacetsRefinements,
+    facetsRefinements,
+    hierarchicalFacetsRefinements,
+    numericRefinements,
+  ].some(refinement => Object.keys(refinement).length > 0);
+}
+
+// A context rule must consist only of alphanumeric characters, hyphens, and underscores.
+// See https://www.algolia.com/doc/guides/managing-results/refine-results/merchandising-and-promoting/in-depth/implementing-query-rules/#context
+function escapeRuleContext(ruleName: string) {
+  return ruleName.replace(/[^a-z0-9-_]+/gi, '_');
+}
+
+function getRuleContextsFromTrackedFilters({
+  helper,
+  sharedHelperState,
+  trackedFilters,
+}: {
+  helper: Helper;
+  sharedHelperState: HelperState;
+  trackedFilters: ParamTrackedFilters;
+}) {
+  const ruleContexts = Object.keys(trackedFilters).reduce<string[]>(
+    (facets, facetName) => {
+      const facetRefinements: Array<string | number> = getRefinements(
+        helper.lastResults || {},
+        sharedHelperState
+      )
+        .filter(
+          (refinement: InternalRefinement) =>
+            refinement.attributeName === facetName
+        )
+        .map(
+          (refinement: InternalRefinement) =>
+            (refinement as InternalNumericRefinement).numericValue ||
```

Comment:
Then we would only process strings, which could be misleading since numeric refinements are numbers:

```
{
  numericRefinements: {
    price: {
      '<=': [500, 400],
      '>=': [100],
    },
  }
}
```

Don't you think?

Reply:
Not sure I understood your point, the only difference is that the return type of the map is an array of string.

Reply:
Yes, which means that the `trackedFilters` functions will get called with strings, even for numeric refinements.

Example:

```js
queryRuleContext({
  price: (values: string[]) => values, // note that prices are now strings
})
```

I think it's a bit misleading because in this example prices are numeric in the records. WDYT?

Reply:
You can keep as is. Note that values are stringified anyway once we send them.